File: frontend/src/api/muestreo.ts
import api from './axios'

export interface MuestreoCreate {
    intento: number
    peso_humedo: number
    peso_seco: number
    observaciones?: string | null
    fecha_muestreo?: string | null
}

export interface MuestreoOut {
    id: number
    lote_id: number
    intento: number
    peso_humedo: number
    peso_seco: number
    porcentaje_humedad: number
    tms_calculado: number
    creado_en: string
    creado_por: number
    observaciones?: string | null
}

export interface MapeoCIPOut {
    id: number
    lote_id: number
    codigo_cip: string
    laboratorio?: string | null
    tipo_muestra?: string | null
    tiene_analisis_ley?: boolean
    tiene_analisis_recuperacion?: boolean
}

export interface MuestreoOfflineItem {
    offline_id: string
    ip: string
    datos: MuestreoCreate
}

export interface SyncMuestreosResponse {
    resultados: Array<{
        offline_id: string
        server_id: number | null
        error: string | null
    }>
}

export interface SyncCipItem {
    offline_id: string
    ip: string
    codigo_cip: string
    correlativo: number
    laboratorio: string
    tipo_muestra: string
}

export interface SyncCipsResponse {
    resultados: Array<{
        offline_id: string
        server_id: number | null
        error: string | null
    }>
}

export interface LoteMuestreo {
    lote_id: number
    ip: string
    fecha_recepcion: string | null
    fecha_muestreo: string | null
    peso_neto: number
    sacos: number | null
    proveedor_razon_social: string
    estado_muestreo: 'PENDIENTE' | 'COMPLETADO'
    cantidad_intentos_previos: number
    tiene_humedad: boolean
    etiquetado: boolean
    fecha_ingreso_prueba: string | null
    sla_config: {
        h_min: number
        h_max: number
    }
    pendiente_sla: boolean
}

export const muestreoApi = {
    /**
     * Registra un muestreo individual estando online.
     */
    async registrarMuestreo(ipLote: string, datos: MuestreoCreate): Promise<MuestreoOut> {
        const response = await api.post<MuestreoOut>(`/muestreo/lotes/${ipLote}`, datos)
        return response.data
    },

    async actualizarMuestreo(id: number, datos: Partial<MuestreoCreate>): Promise<MuestreoOut> {
        const response = await api.patch<MuestreoOut>(`/muestreo/${id}`, datos)
        return response.data
    },

    async listarMuestreosPorLote(ipLote: string): Promise<MuestreoOut[]> {
        const response = await api.get<MuestreoOut[]>(`/muestreo/lotes/${ipLote}/muestreos`)
        return response.data
    },

    /**
     * Registra múltiples muestreos en lote estando online.
     */
    async registrarMuestreoBatch(ipLote: string, datosList: MuestreoCreate[]): Promise<MuestreoOut[]> {
        const response = await api.post<MuestreoOut[]>(`/muestreo/lotes/${ipLote}/batch`, datosList)
        return response.data
    },

    /**
     * Sincroniza un bloque de muestreos guardados offline.
     */
    async syncBatch(muestreos: MuestreoOfflineItem[]): Promise<SyncMuestreosResponse> {
        const response = await api.post<SyncMuestreosResponse>('/muestreo/sync', { muestreos })
        return response.data
    },

    /**
     * Genera los códigos CIP (Muestreo Ciego) para el laboratorio. (y minero y testigo)
     */
    async generarCips(ipLote: string, cantidad: number = 5): Promise<MapeoCIPOut[]> {
        const response = await api.post<MapeoCIPOut[]>(`/muestreo/lotes/${ipLote}/etiquetas`, {
            cantidad,
            laboratorio: 'Por definir'
        })
        return response.data
    },

    /**
     * Obtiene la lista de lotes que requieren muestreo.
     */
    async obtenerLotes(): Promise<LoteMuestreo[]> {
        // Ajusta la ruta según tu backend real
        const response = await api.get<LoteMuestreo[]>('/muestreo/lotes')
        return response.data
    },

    /**
     * Obtiene las etiquetas CIP generadas para un lote específico.
     */
    async obtenerEtiquetas(ipLote: string): Promise<MapeoCIPOut[]> {
        const response = await api.get<MapeoCIPOut[]>(`/muestreo/lotes/${ipLote}/etiquetas`)
        return response.data
    },

    async actualizarLaboratorioCip(cipId: number, laboratorio: string): Promise<MapeoCIPOut> {
        const response = await api.patch<MapeoCIPOut>(`/muestreo/cips/${cipId}/laboratorio`, { laboratorio })
        return response.data
    },

    async listarLaboratorios(): Promise<string[]> {
        const response = await api.get<string[]>('/muestreo/labs')
        return response.data
    },

    /**
     * Sincroniza los CIPs generados offline al reconectar.
     * Idempotente: el servidor no duplica si el código ya existe.
     */
    async syncCips(cips: SyncCipItem[]): Promise<SyncCipsResponse> {
        const response = await api.post<SyncCipsResponse>('/muestreo/sync-cips', { cips })
        return response.data
    },
}

File: backend/app/schemas/muestreo.py
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, Field


# ==========================================
# MUESTREO (HUMEDAD)
# ==========================================
class MuestreoCreate(BaseModel):
    intento: int = Field(..., ge=1, le=3, description="Número de intento (1, 2 o 3)")
    peso_humedo: Decimal = Field(..., gt=0)
    peso_seco: Decimal = Field(..., gt=0)
    observaciones: str | None = None
    fecha_muestreo: datetime | None = None


class MuestreoUpdate(BaseModel):
    peso_humedo: Decimal = Field(..., gt=0)
    peso_seco: Decimal = Field(..., gt=0)
    observaciones: str | None = None


class MuestreoOut(BaseModel):
    id: int
    lote_id: int
    intento: int
    peso_humedo: Decimal
    peso_seco: Decimal
    porcentaje_humedad: Decimal
    tms_calculado: Decimal
    creado_en: datetime
    creado_por: int
    observaciones: str | None = None

    model_config = {"from_attributes": True}


# ==========================================
# CÓDIGOS CIP
# ==========================================
class GenerarCipsRequest(BaseModel):
    cantidad: int = Field(default=2, ge=1, le=5, description="Cantidad de bolsas para laboratorio")
    laboratorio: str | None = "Por definir"


class MapeoCIPOut(BaseModel):
    id: int
    lote_id: int
    codigo_cip: str
    laboratorio: str | None
    tipo_muestra: str | None
    tiene_analisis_ley: bool = False
    tiene_analisis_recuperacion: bool = False

    model_config = {"from_attributes": True}


class ActualizarLabCIPRequest(BaseModel):
    laboratorio: str


# ==========================================
# SYNC OFFLINE
# ==========================================
class MuestreoOfflineItem(BaseModel):
    offline_id: str
    ip: str
    datos: MuestreoCreate


class SyncMuestreosRequest(BaseModel):
    muestreos: list[MuestreoOfflineItem]


class SyncResult(BaseModel):
    offline_id: str
    server_id: int | None = None
    error: str | None = None


class SyncMuestreosResponse(BaseModel):
    resultados: list[SyncResult]


# ==========================================
# SYNC CIPs OFFLINE
# ==========================================
class SyncCipItem(BaseModel):
    offline_id: str
    ip: str
    codigo_cip: str
    correlativo: int
    laboratorio: str
    tipo_muestra: str = "Laboratorio"


class SyncCipsRequest(BaseModel):
    cips: list[SyncCipItem]


class SyncCipResult(BaseModel):
    offline_id: str
    server_id: int | None = None
    error: str | None = None


class SyncCipsResponse(BaseModel):
    resultados: list[SyncCipResult]
